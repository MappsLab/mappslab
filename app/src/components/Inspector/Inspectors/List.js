// @flow
import * as React from 'react'
import styled from 'styled-components'
import { Header2, Header5 } from 'Components/Text'
import ItemIcon from '../ItemIcon'

const ListTitle = styled(Header2)`
	${({ theme }) => `
		color: ${theme.color.primary.light}
	`}
`

const ListWrapper = styled.div`
	${({ theme }) => `
		margin-top: ${theme.layout.spacing.quadruple};
	`}
`

const ListItem = styled.div`
	${({ theme }) => `
		display: flex;
		justify-content: space-between;
		padding: ${theme.layout.spacing.half} ${theme.layout.spacing.half};
		border-top: 1px solid ${theme.color.middleGray};

		&:hover {
			background-color: ${theme.color.xLightGray};
		}

		&:last-child {
			border-bottom: 1px solid ${theme.color.middleGray};
		}
	`}
`

const ItemTitle = styled(Header5)`
	${({ theme }) => `
		font-weight: ${theme.font.weight.semi};
	`}
`

const ItemInfo = styled(Header5)`
	${({ theme }) => `
		color: ${theme.color.middleGray};
	`}
`

export type ListItemType = {
	key: string,
	title: string,
	info: Array<string>,
	onClick: () => void,
}

type Props = {
	items: Array<ListItemType>,
	title: string,
	type: string,
}

/**
 * List
 */

const List = ({ items, title, type }: Props) => (
	<ListWrapper>
		<ListTitle>
			<ItemIcon type={type} />
			{title}
		</ListTitle>
		{items.map(({ key, title: itemTitle, info, onClick }) => (
			<ListItem key={key} onClick={onClick}>
				<ItemTitle>{itemTitle}</ItemTitle>
				{info.map((i) => (
					<ItemInfo key={i}>{i}</ItemInfo>
				))}
			</ListItem>
		))}
	</ListWrapper>
)

export default List
