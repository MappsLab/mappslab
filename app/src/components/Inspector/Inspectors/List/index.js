// @flow

import WrappedList from './List'

export default WrappedList
