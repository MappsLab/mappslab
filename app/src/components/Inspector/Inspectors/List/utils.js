// @flow
import type { ClassroomType, UserType, MapType } from 'Types'
// import type { InspectorItem } from '../../InspectorProvider'

type NodeArray = Array<ClassroomType | UserType | MapType>

export type SearchForList = (any) => Promise<NodeArray> | NodeArray

export type ListItemHandler = (ClassroomType | UserType | MapType) => void

export type ListItem = {
	node: ClassroomType | UserType | MapType,
	key: string,
	onClick: () => void,
}

// const nodeToInspectorItem = (item: ClassroomType | UserType | MapType): InspectorItem => ({
// 	uid: item.uid,
// 	type: item.__typename,
// 	title: item.__typename === 'User' ? item.name : item.title,
// })

export const nodeToListItem = (node: ClassroomType | UserType | MapType, handler: ListItemHandler): ListItem => {
	// $FlowFixMe TODO: make this detect types
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
