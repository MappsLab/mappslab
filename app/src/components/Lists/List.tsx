import * as React from 'react'
import { Node } from '../../types-ts'
import ItemIcon from 'Components/ItemIcon'
import { ListAddEntry } from './ListAddEntry'
import { ButtonConfig } from './ButtonDropdown'
import { ListItem } from './ListItem'
import { ListTitle, ListWrapper, ListItems } from './styled'
import {
	CreateNewFn,
	SearchForList,
	nodeToListItem,
	ListItemHandler,
} from './utils'

/**
 * List
 */

interface Props<NodeType extends Node> {
	items: NodeType[]
	title: string
	type: string
	viewerCanAdd?: boolean
	addLabel?: string
	search?: SearchForList
	create?: CreateNewFn
	onSearchResultClick?: ListItemHandler
	searchResults?: NodeType[]
	onItemClick?: ListItemHandler
	buttons?: ButtonConfig<NodeType>[]
}

const defaultAddLabel = 'Add'
const defaultOnItemClick = async () => {}

export const List = <NodeType extends Node>({
	items,
	title,
	type,
	onItemClick,
	viewerCanAdd,
	addLabel,
	search,
	create,
	onSearchResultClick,
	searchResults,
	buttons,
}: Props<NodeType>) => {
	const itemToListItem = (node: NodeType) =>
		nodeToListItem(node, onItemClick || defaultOnItemClick)
	if (viewerCanAdd && (!search || !onSearchResultClick || !create))
		throw new Error(
			'You must provide `search`, `onSearchResultClick`, and `create` functions',
		)
	return (
		<ListWrapper>
			<ListTitle>
				<ItemIcon type={type} />
				{title}
			</ListTitle>
			<ListItems>
				{items.map(itemToListItem).map((item) => (
					<ListItem key={item.key} {...item} buttons={buttons} />
				))}
			</ListItems>
			{viewerCanAdd && search && onSearchResultClick && create && (
				<ListAddEntry
					search={search}
					type={type}
					create={create}
					searchResults={searchResults}
					onSearchResultClick={onSearchResultClick}
					searchName={type.replace(/s$/, '')}
					addLabel={addLabel || defaultAddLabel}
				/>
			)}
		</ListWrapper>
	)
}

List.defaultProps = {
	viewerCanAdd: false,
	addLabel: defaultAddLabel,
	search: undefined,
	create: undefined,
	onSearchResultClick: undefined,
	searchResults: [],
	onItemClick: async () => {},
}
