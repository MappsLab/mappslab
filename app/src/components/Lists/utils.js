// @flow
import type { Node, Variables } from 'Types'
// import type { InspectorItem } from '../../InspectorProvider'

// type NodeArray = Array<ClassroomType | UserType | MapType>

export type SearchForList = (any) => Promise<void> | void

export type ListItemHandler = (Node) => void

export type ListItem = {
	node: Node,
	onClick: () => void,
}

export type ListOfTypeBaseProps<Type> = {
	title: string,
	items: Array<Type>,
	viewerCanAdd: boolean,
	update: (Type) => void,
	onItemClick: ListItemHandler,
}

export type ListOfTypeProps<Type> = ListOfTypeBaseProps<Type> & {
	searchQuery: (Variables) => Promise<void>,
	searchResults: Array<Type>,
}

export const nodeToListItem = (node: Node, handler: ListItemHandler): ListItem => {
	// $FlowFixMe TODO: make this detect types
	const { uid } = node
	return {
		key: uid,
		node,
		info: [],
		onClick: () => handler(node),
	}
}
