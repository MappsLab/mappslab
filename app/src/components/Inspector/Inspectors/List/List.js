// @flow
import * as React from 'react'
import styled from 'styled-components'
import { Header2 } from 'Components/Text'
import type { ClassroomType, UserType, MapType } from 'Types'
import ItemIcon from '../../ItemIcon'
import { InspectorConsumer } from '../../InspectorProvider'
import type { InspectItem } from '../../InspectorProvider'
import ListAddEntry from './ListAddEntry'
import ListItem from './ListItem'
import type { SearchForList, ListItemHandler } from './utils'
import { nodeToListItem } from './utils'

const ListTitle = styled(Header2)`
	${({ theme }) => `
		color: ${theme.color.primary.light}
	`}
`

const ListWrapper = styled.div`
	${({ theme }) => `
		margin-top: ${theme.layout.spacing.quadruple};
	`}
`

/**
 * List
 */

type BaseProps = {
	items: Array<ClassroomType | UserType | MapType>,
	title: string,
	type: string,
	viewerCanAdd?: boolean,
	addLabel?: string,
	search?: SearchForList,
	onSearchResultClick?: ListItemHandler,
}

type Props = BaseProps & {
	inspectItem: InspectItem,
}

const defaultAddLabel = 'Add'

export const List = ({ items, title, type, inspectItem, viewerCanAdd, addLabel, search, onSearchResultClick }: Props) => {
	const itemToListItem = (node) => nodeToListItem(node, inspectItem)
	if (viewerCanAdd && (!search || !onSearchResultClick))
		throw new Error('You must provide `search` and `onSearchResultClick` functions')
	return (
		<ListWrapper>
			<ListTitle>
				<ItemIcon type={type} />
				{title}
			</ListTitle>
			{items.map(itemToListItem).map((item) => (
				<ListItem key={item.key} {...item} />
			))}
			{viewerCanAdd && search && onSearchResultClick && (
				<ListAddEntry
					search={search}
					onSearchResultClick={onSearchResultClick}
					searchName={type.replace(/s$/, '')}
					addLabel={addLabel || defaultAddLabel}
				/>
			)}
		</ListWrapper>
	)
}

List.defaultProps = {
	viewerCanAdd: false,
	addLabel: defaultAddLabel,
	search: undefined,
	onSearchResultClick: undefined,
}

export default (baseProps: BaseProps) => (
	<InspectorConsumer>{({ inspectItem }) => <List {...baseProps} inspectItem={inspectItem} />}</InspectorConsumer>
)
