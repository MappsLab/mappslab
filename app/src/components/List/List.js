// @flow
import * as React from 'react'
import styled from 'styled-components'
import { Header2 } from 'Components/Text'
import type { ClassroomType, UserType, MapType } from 'Types'
import ItemIcon from 'Components/ItemIcon'
import ListAddEntry from './ListAddEntry'
import ListItem from './ListItem'
import type { SearchForList, ListItemHandler } from './utils'
import { nodeToListItem } from './utils'

const ListTitle = styled(Header2)`
	${({ theme }) => `
		color: ${theme.color.primary.light}
	`}
`

const ListWrapper = styled.div`
	${({ theme }) => `
		margin-top: ${theme.layout.spacing.quadruple};
	`}
`

/**
 * List
 */

type Props = {
	items: Array<ClassroomType | UserType | MapType>,
	title: string,
	type: string,
	viewerCanAdd?: boolean,
	addLabel?: string,
	search?: SearchForList,
	onSearchResultClick?: ListItemHandler,
	searchResults?: Array<ClassroomType | UserType | MapType>,
	onItemClick: ListItemHandler,
}

const defaultAddLabel = 'Add'

export const List = ({
	items,
	title,
	type,
	onItemClick,
	viewerCanAdd,
	addLabel,
	search,
	onSearchResultClick,
	searchResults,
}: Props) => {
	const itemToListItem = (node) => nodeToListItem(node, onItemClick)
	if (viewerCanAdd && (!search || !onSearchResultClick))
		throw new Error('You must provide `search` and `onSearchResultClick` functions')
	return (
		<ListWrapper>
			<ListTitle>
				<ItemIcon type={type} />
				{title}
			</ListTitle>
			{items.map(itemToListItem).map((item) => (
				<ListItem key={item.node.uid} {...item} />
			))}
			{viewerCanAdd && search && onSearchResultClick && (
				<ListAddEntry
					search={search}
					searchResults={searchResults}
					onSearchResultClick={onSearchResultClick}
					searchName={type.replace(/s$/, '')}
					addLabel={addLabel || defaultAddLabel}
				/>
			)}
		</ListWrapper>
	)
}

List.defaultProps = {
	viewerCanAdd: false,
	addLabel: defaultAddLabel,
	search: undefined,
	onSearchResultClick: undefined,
	searchResults: [],
}

export default List

// export default (baseProps: BaseProps) => (
// 	<InspectorConsumer>{({ inspectItem }) => <List {...baseProps} inspectItem={inspectItem} />}</InspectorConsumer>
// )
