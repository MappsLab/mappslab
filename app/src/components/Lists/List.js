// @flow
import * as React from 'react'
import type { Node } from 'Types'
import ItemIcon from 'Components/ItemIcon'
import ListAddEntry from './ListAddEntry'
import ListItem from './ListItem'
import { ListTitle, ListWrapper, ListItems } from './styled'
import type { CreateNewFn, SearchForList, ListItemHandler } from './utils'
import { nodeToListItem } from './utils'

/**
 * List
 */

type Props = {
	items: Array<Node>,
	title: string,
	type: string,
	viewerCanAdd?: boolean,
	addLabel?: string,
	search?: SearchForList,
	create?: CreateNewFn,
	onSearchResultClick?: ListItemHandler,
	searchResults?: Array<Node>,
	onItemClick?: ListItemHandler,
}

const defaultAddLabel = 'Add'
const defaultOnItemClick = async () => {}

const List = ({
	items,
	title,
	type,
	onItemClick,
	viewerCanAdd,
	addLabel,
	search,
	create,
	onSearchResultClick,
	searchResults,
}: Props) => {
	const itemToListItem = (node) => nodeToListItem(node, onItemClick || defaultOnItemClick)
	if (viewerCanAdd && (!search || !onSearchResultClick || !create))
		throw new Error('You must provide `search`, `onSearchResultClick`, and `create` functions')
	return (
		<ListWrapper>
			<ListTitle>
				<ItemIcon type={type} />
				{title}
			</ListTitle>
			<ListItems>
				{items.map(itemToListItem).map((item) => (
					<ListItem key={item.key} {...item} />
				))}
			</ListItems>
			{viewerCanAdd && search && onSearchResultClick && create && (
				<ListAddEntry
					search={search}
					type={type}
					create={create}
					searchResults={searchResults}
					onSearchResultClick={onSearchResultClick}
					searchName={type.replace(/s$/, '')}
					addLabel={addLabel || defaultAddLabel}
				/>
			)}
		</ListWrapper>
	)
}

List.defaultProps = {
	viewerCanAdd: false,
	addLabel: defaultAddLabel,
	search: undefined,
	create: undefined,
	onSearchResultClick: undefined,
	searchResults: [],
	onItemClick: async () => {},
}

export default List
