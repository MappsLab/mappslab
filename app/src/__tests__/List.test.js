import * as React from 'react'
import { render, mockServer } from 'Jest/utils'
import { fireEvent, wait } from 'react-testing-library'
import { act } from 'react-dom/test-utils'
/* import the bare component so we can inject dependencies instead of dealing with context */
import { List } from 'Components/Inspector/Inspectors/List/List'

const types = ['classrooms', 'teachers', 'students']

const noop = () => {}

const getTypes = async () => {
	const [classrooms, students, teachers] = await Promise.all([
		mockServer.getClassrooms(),
		mockServer.getStudents(),
		mockServer.getTeachers(),
		// mockServer.getMaps(),
	])
	return { classrooms, students, teachers }
}

describe('List Component', () => {
	/* run the suite for each type of entity */

	types.forEach((type) => {
		it(`should list items [${type}]`, async () => {
			const result = await getTypes()
			const items = result[type]
			const inspectItem = jest.fn()

			const { queryByText } = render(<List inspectItem={inspectItem} title="My List" items={items} />)
			items.forEach((item) => {
				const text = item.title || item.name
				expect(queryByText(text)).toBeTruthy()
			})
		})
	})

	types.forEach((type) => {
		it(`should call inspectItem with their info when clicked [${type}]`, async () => {
			const result = await getTypes()
			const items = result[type]
			const inspectItem = jest.fn()
			const { getByText } = render(<List inspectItem={inspectItem} type={type} title="My List" items={items} />)

			/* click on each item */
			items.forEach((item) => {
				const { uid, title, name, __typename } = item
				const itemTitle = title || name
				const itemButton = getByText(itemTitle)
				fireEvent.click(itemButton)
				if (__typename === 'User') {
					expect(inspectItem).toHaveBeenCalledWith({ uid, name: itemTitle, __typename })
				} else {
					expect(inspectItem).toHaveBeenCalledWith({ uid, title: itemTitle, __typename })
				}
			})
		})
	})

	it('should by default not display the + Add button', async () => {
		const inspectItem = jest.fn()
		const { classrooms } = await getTypes()
		const { queryByText } = render(<List inspectItem={inspectItem} type="Classrooms" title="My Classrooms" items={classrooms} />)
		expect(queryByText('+ Add')).toBeFalsy()
	})

	it('should allow users to make new associations when given `viewerCanAdd === true`', async () => {
		const { classrooms } = await getTypes()
		const { getByText } = render(
			<List
				inspectItem={noop}
				type="Classrooms"
				search={noop}
				onSearchResultClick={noop}
				title="My Classrooms"
				viewerCanAdd
				items={classrooms}
			/>,
		)
		expect(getByText('+ Add')).toBeTruthy()
	})

	it('should display custom `addLabel` text', async () => {
		const { classrooms } = await getTypes()
		const { getByText } = render(
			<List
				inspectItem={noop}
				type="Classrooms"
				title="My Classrooms"
				viewerCanAdd
				search={noop}
				onSearchResultClick={noop}
				addLabel="Add Classroom"
				items={classrooms}
			/>,
		)
		expect(getByText('+ Add Classroom')).toBeTruthy()
	})

	it('should search for items, display them, and call the `addNew()` function', async () => {
		const { classrooms } = await getTypes()
		const [class1, class2, ...otherClassrooms] = classrooms
		const search = jest.fn(async () => [class1, class2])
		const onSearchResultClick = jest.fn()
		const { container, getByText, queryByText } = render(
			<List
				inspectItem={noop}
				type="Classrooms"
				title="My Classrooms"
				viewerCanAdd
				search={search}
				onSearchResultClick={onSearchResultClick}
				items={otherClassrooms}
			/>,
		)
		let addButton = getByText('+ Add')
		/* expect a search input to appear after clicking the add button */
		fireEvent.click(addButton)
		const searchInput = container.querySelector('input[id="searchInput"]')
		expect(searchInput).toBeTruthy()

		await wait()
		act(() => {
			fireEvent.change(searchInput, { target: { value: 'So' } })
		})

		/* expect the search fn to be called after the input changes */
		expect(search.mock.calls[0][0]).toEqual('So')
		const val1 = await search.mock.results[0].value
		expect(val1).toEqual([class1, class2])

		await wait()

		/* expect search results to show up as new list items */
		const class1Result = getByText(class1.title)
		const class2Result = getByText(class2.title)
		expect(class1Result).toBeTruthy()
		expect(class2Result).toBeTruthy()

		act(() => {
			fireEvent.click(class1Result)
		})

		/* Expect the click handler to have been called */
		const { uid, __typename, title } = class1
		expect(onSearchResultClick.mock.calls[0][0]).toEqual({ uid, __typename, title })

		/* Expect the form to have been reset */
		addButton = getByText('+ Add')
		expect(addButton).toBeTruthy()
		expect(queryByText(class2.title)).toBeFalsy()
	})

	it('should throw an error if `viewerCanAdd` is supplied without a `search` function', async () => {
		const { classrooms } = await getTypes()

		expect(() => {
			render(
				<List
					inspectItem={noop}
					type="Classrooms"
					title="My Classrooms"
					viewerCanAdd
					onSearchResultClick={noop}
					items={classrooms}
				/>,
			)
		}).toThrowErrorMatchingSnapshot()

		expect(() => {
			render(<List inspectItem={noop} type="Classrooms" title="My Classrooms" viewerCanAdd search={noop} items={classrooms} />)
		}).toThrowErrorMatchingSnapshot()

		/* This should not throw */
		render(
			<List
				inspectItem={noop}
				type="Classrooms"
				title="My Classrooms"
				viewerCanAdd
				search={noop}
				onSearchResultClick={noop}
				items={classrooms}
			/>,
		)
	})
})
