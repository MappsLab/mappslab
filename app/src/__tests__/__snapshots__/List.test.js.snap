// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Component should throw an error if \`viewerCanAdd\` is supplied without a \`search\` function 1`] = `"You must provide \`search\` and \`onSearchResultClick\` functions"`;

exports[`List Component should throw an error if \`viewerCanAdd\` is supplied without a \`search\` function 2`] = `"You must provide \`search\` and \`onSearchResultClick\` functions"`;
